### ISSUE TITLE
Incorrect password with docker_registry causes infinite retries & server lockout

### ISSUE THREAD

**Issue by mrdziuban** [bflad/chef-docker#386]
When the password passed to the `docker_registry` resource is incorrect, there seems to be an infinite number of retries to log into the registry. After 10 tries, the server locks me out of attempting more logins for 1 hour, but the retries continue. Here's an example:

```
docker_service 'default' do
  action [:create, :start]
end

docker_registry 'https://index.docker.io/v1/' do
  username 'mrdziuban'
  password 'INVALID_PASSWORD'
  email 'email@email.com'
end
```

This is what `/var/log/upstart/docker.log` looks like:

```
Wrong login/password, please try again
time="2015-08-21T17:33:24Z" level=info msg="-job auth() = ERR (1)" 
time="2015-08-21T17:33:24Z" level=error msg="Handler for POST /auth returned error: Wrong login/password, please try again" 
time="2015-08-21T17:33:24Z" level=error msg="HTTP Error: statusCode=401 Wrong login/password, please try again" 
time="2015-08-21T17:33:24Z" level=info msg="POST /v1.16/auth" 
time="2015-08-21T17:33:24Z" level=info msg="+job auth()" 
time="2015-08-21T17:33:24Z" level=info msg="+job resolve_index(https://index.docker.io/v1/)" 
time="2015-08-21T17:33:24Z" level=info msg="-job resolve_index(https://index.docker.io/v1/) = OK (0)" 
time="2015-08-21T17:33:25Z" level=error msg="unable to login against registry endpoint https://index.docker.io/v1/: Wrong login/password, please try again"

...

time="2015-08-21T17:33:41Z" level=info msg="POST /v1.16/auth" 
time="2015-08-21T17:33:41Z" level=info msg="+job auth()" 
time="2015-08-21T17:33:41Z" level=info msg="+job resolve_index(https://index.docker.io/v1/)" 
time="2015-08-21T17:33:41Z" level=info msg="-job resolve_index(https://index.docker.io/v1/) = OK (0)" 
time="2015-08-21T17:33:41Z" level=error msg="unable to login against registry endpoint https://index.docker.io/v1/: Login: Too many failed login attempts. You have been locked out for one hour. Try again later or contact support. (Code: 429; Headers: map[X-Frame-Options:[SAMEORIGIN] Strict-Transport-Security:[max-age=31536000] Server:[nginx/1.6.2] Date:[Fri, 21 Aug 2015 17:33:41 GMT] Content-Type:[application/json] Vary:[Cookie]])" 
Login: Too many failed login attempts. You have been locked out for one hour. Try again later or contact support. (Code: 429; Headers: map[X-Frame-Options:[SAMEORIGIN] Strict-Transport-Security:[max-age=31536000] Server:[nginx/1.6.2] Date:[Fri, 21 Aug 2015 17:33:41 GMT] Content-Type:[application/json] Vary:[Cookie]])
time="2015-08-21T17:33:41Z" level=info msg="-job auth() = ERR (1)" 
time="2015-08-21T17:33:41Z" level=error msg="Handler for POST /auth returned error: Login: Too many failed login attempts. You have been locked out for one hour. Try again later or contact support. (Code: 429; Headers: map[X-Frame-Options:[SAMEORIGIN] Strict-Transport-Security:[max-age=31536000] Server:[nginx/1.6.2] Date:[Fri, 21 Aug 2015 17:33:41 GMT] Content-Type:[application/json] Vary:[Cookie]])" 
time="2015-08-21T17:33:41Z" level=error msg="HTTP Error: statusCode=500 Login: Too many failed login attempts. You have been locked out for one hour. Try again later or contact support. (Code: 429; Headers: map[X-Frame-Options:[SAMEORIGIN] Strict-Transport-Security:[max-age=31536000] Server:[nginx/1.6.2] Date:[Fri, 21 Aug 2015 17:33:41 GMT] Content-Type:[application/json] Vary:[Cookie]])"
```

This is happening at least on versions 1.0.7 and 1.0.8. I've confirmed on versions 0.37.0 and 0.43.0 that this is not an issue.

**Comment by mrdziuban** [comment]
@chasebolt Thanks for opening up that PR. I'm not sure if that will solve my issue, however. I've tracked down the issue to this line: https://github.com/bflad/chef-docker/blob/v1.0.8/libraries/provider_docker_registry.rb#L19.

Although I expect `tries` to be set to 3, due to [this default value](https://github.com/bflad/chef-docker/blob/v1.0.8/libraries/resource_docker_registry.rb#L13), it's set to 0 when starting, so `(tries -= 1)` will never be `zero?`. So it seems like the default value might not be working correctly?

**Comment by rpazyaquian** [comment]
To expand on the issue, here's an example of it in action.

Assuming that you have a default recipe like this:

```
docker_service 'default' do
  action [:create, :start]
end

docker_registry 'https://index.docker.io/v1/' do
  username 'username'
  password 'invalid_password'
  email 'some@email.address'
end
```

And there is some debugging code in `provider_docker_registry.rb`:

```
class Chef
  class Provider
    class DockerRegistry < Chef::Provider::LWRPBase
      provides :docker_registry if Chef::Provider.respond_to?(:provides)

      action :login do
        puts "\nnew_resource.retries: #{new_resource.retries}"
        [ ...rest of code ]
```

Then you get an output like this:

```
         * docker_service[default] action create
           * remote_file[/usr/bin/docker] action create (up to date)
            (up to date)
         * docker_service[default] action start
           * template[/etc/default/docker] action create (up to date)
           * template[/etc/init/docker.conf] action create (up to date)
           * service[docker] action start (up to date)
            (up to date)
         * docker_registry[https://index.docker.io/v1/] action login
       new_resource.retries: 0
       [ ...hangs indefinitely ]
```

Where `new_resource.retries` is incorrectly set to `3`, not `0`.

However, if you explicitly specify the number of retries:

```
docker_service 'default' do
  action [:create, :start]
end

docker_registry 'https://index.docker.io/v1/' do
  username 'username'
  password 'invalid_password'
  email 'some@email.address'
  retries 1
end
```

Then you get an output like this:

```
         * docker_service[default] action create
           * remote_file[/usr/bin/docker] action create (up to date)
            (up to date)
         * docker_service[default] action start
           * template[/etc/default/docker] action create (up to date)
           * template[/etc/init/docker.conf] action create (up to date)
           * service[docker] action start (up to date)
            (up to date)
         * docker_registry[https://index.docker.io/v1/] action login
       new_resource.retries: 1
       [ ...successful converge ]
```

I believe that `new_resource.retries` is not actually set to `3` by default, but to `0` somehow. This might depend on how the `new_resource` object is created and how the `retries` attribute is set.

**Comment by chasebolt** [comment]
ok i do see this issue, i assumed it was upstart from the log output and upstart also retries 10 times; doh.

**Comment by mrdziuban** [comment]
Can we reopen this issue please since #387 does not address the default value not working properly?

**Comment by chasebolt** [comment]
the new PR should take care of this issue.

**Comment by someara** [comment]
This should be fixed in 1.0.11

**Comment by mrdziuban** [comment]
Thanks @someara, the fix worked for me. I'm now noticing that the converge does not fail if the login is unsuccessful. Moving conversation to #398.
